File: packages/parallax_cli/src/commands/create.ts
// Create command and all its subcommands (image, video, audio).
// Action handlers follow: validate → resolveModelsDir → buildArgs → spawnPipeline.
// All model-specific special cases are encapsulated in the models/ builders.

import { existsSync } from "fs";
import { Command } from "commander";
import { readConfig } from "../config";
import { spawnPipeline, runAsync } from "../runner";
import { resolveModelsDir } from "../utils";
import { getModels, getScript, getModelConfig, getModelDefaults, type ModelDefaults } from "../models/registry";
import { buildArgs as buildImageArgs, type ImageOpts } from "../models/image";
import { buildArgs as buildVideoArgs, type VideoOpts } from "../models/video";
import { buildArgs as buildAudioArgs, type AudioOpts } from "../models/audio";
import { modelsFooter, validateModel, notImplemented, resolveParam } from "../utils";

// Columns per media type: [key in ModelDefaults, display label]
const DEFAULTS_COLUMNS: Record<string, Array<[keyof ModelDefaults, string]>> = {
  video: [["width","width"],["height","height"],["length","length"],["fps","fps"],["steps","steps"],["cfg","cfg"]],
  image: [["width","width"],["height","height"],["steps","steps"],["cfg","cfg"]],
  audio: [["length","duration"],["steps","steps"],["cfg","cfg"],["bpm","bpm"]],
};

export function buildDefaultsTable(media: string): string {
  const cols = DEFAULTS_COLUMNS[media];
  if (!cols) return "";
  const models = getModels("create", media).filter(m => getModelDefaults(media, m) !== undefined);
  if (models.length === 0) return "";

  const headers = ["model", ...cols.map(([, label]) => label)];
  const rows: string[][] = models.map(model => {
    const d = getModelDefaults(media, model)!;
    return [model, ...cols.map(([key]) => d[key] != null ? String(d[key]) : "—")];
  });

  const widths = headers.map((h, i) => Math.max(h.length, ...rows.map(r => r[i].length)));
  const pad = (s: string, w: number) => s.padEnd(w);
  const headerRow = "  " + headers.map((h, i) => pad(h, widths[i])).join("  ");
  const sepRow = "  " + "─".repeat(headerRow.length - 2);
  const dataRows = rows.map(row => "  " + row.map((cell, i) => pad(cell, widths[i])).join("  "));

  return "\nDefaults per model:\n\n" + headerRow + "\n" + sepRow + "\n" + dataRows.join("\n");
}

export function registerCreate(program: Command): void {
  const create = program
    .command("create")
    .description("Generate media from a text prompt")
    .usage("<media> [options]");

  create
    .command("image")
    .description("Generate an image from a text prompt")
    .requiredOption("--model <name>", `Model to use (choices: ${getModels("create", "image").join(", ")})`)
    .requiredOption("--prompt <text>", "Text prompt describing the image to generate")
    .option("--negative-prompt <text>", "Negative prompt (what to avoid in the output)")
    .option("--width <pixels>", "Image width in pixels")
    .option("--height <pixels>", "Image height in pixels")
    .option("--steps <n>", "Number of sampling steps")
    .option("--cfg <value>", "CFG guidance scale")
    .option("--seed <n>", "Random seed for reproducibility")
    .option("--output <path>", "Output file path", "output.png")
    .option("--models-dir <path>", "Models directory (overrides PYCOMFY_MODELS_DIR)")
    .option("--async", "Queue job and return a job ID immediately (non-blocking)")
    .addHelpText("after", modelsFooter("create", "image"))
    .addHelpText("after", buildDefaultsTable("image"))
    .action(async (opts) => {
      validateModel("create", "image", opts.model);
      const script = getScript("create", "image", opts.model);
      if (!script) notImplemented("create", "image", opts.model);
      const modelsDir = resolveModelsDir(opts.modelsDir);
      const defaults = getModelDefaults("image", opts.model);
      const imageOpts: ImageOpts = {
        model:          opts.model,
        prompt:         opts.prompt,
        negativePrompt: opts.negativePrompt,
        width:          resolveParam(opts.width,  defaults?.width,  "width"),
        height:         resolveParam(opts.height, defaults?.height, "height"),
        steps:          resolveParam(opts.steps,  defaults?.steps,  "steps"),
        cfg:            resolveParam(opts.cfg,    defaults?.cfg,    "cfg"),
        seed:           opts.seed,
        output:         opts.output,
      };
      const args = buildImageArgs(imageOpts, modelsDir);
      if (opts.async) {
        await runAsync("create", "image", opts.model, script, args, readConfig());
      } else {
        await spawnPipeline(script, args, readConfig());
      }
    });

  create
    .command("video")
    .description("Generate a video from a text prompt")
    .requiredOption("--model <name>", `Model to use (choices: ${getModels("create", "video").join(", ")})`)
    .requiredOption("--prompt <text>", "Text prompt describing the video to generate")
    .option("--input <path>", "Input image path for image-to-video (ltx2, ltx23, wan21, wan22)")
    .option("--width <pixels>", "Video width in pixels")
    .option("--height <pixels>", "Video height in pixels")
    .option("--length <frames>", "Number of frames to generate")
    .option("--steps <n>", "Number of sampling steps")
    .option("--cfg <value>", "CFG guidance scale")
    .option("--seed <n>", "Random seed for reproducibility")
    .option("--output <path>", "Output file path", "output.mp4")
    .option("--models-dir <path>", "Models directory (overrides PYCOMFY_MODELS_DIR)")
    .option("--async", "Queue job and return a job ID immediately (non-blocking)")
    .addHelpText("after", modelsFooter("create", "video"))
    .addHelpText("after", buildDefaultsTable("video"))
    .action(async (opts) => {
      validateModel("create", "video", opts.model);
      if (opts.input !== undefined && !existsSync(opts.input)) {
        console.error(`Error: input file not found: ${opts.input}`);
        process.exit(1);
      }
      const modelConfig = getModelConfig("video", opts.model);
      const useI2v = opts.input !== undefined && modelConfig?.i2v !== undefined;
      const script = modelConfig ? (useI2v ? modelConfig.i2v! : modelConfig.t2v) : undefined;
      if (!script) notImplemented("create", "video", opts.model);
      const modelsDir = resolveModelsDir(opts.modelsDir);
      const defaults = getModelDefaults("video", opts.model);
      const videoOpts: VideoOpts = {
        model:  opts.model,
        prompt: opts.prompt,
        input:  opts.input,
        width:  resolveParam(opts.width,  defaults?.width,  "width"),
        height: resolveParam(opts.height, defaults?.height, "height"),
        length: resolveParam(opts.length, defaults?.length, "length"),
        steps:  resolveParam(opts.steps,  defaults?.steps,  "steps"),
        cfg:    resolveParam(opts.cfg,    defaults?.cfg,    "cfg"),
        seed:   opts.seed,
        output: opts.output,
      };
      const args = buildVideoArgs(videoOpts, modelsDir);
      if (opts.async) {
        await runAsync("create", "video", opts.model, script, args, readConfig());
      } else {
        await spawnPipeline(script, args, readConfig());
      }
    });

  create
    .command("audio")
    .description("Generate audio from a text prompt")
    .requiredOption("--model <name>", `Model to use (choices: ${getModels("create", "audio").join(", ")})`)
    .requiredOption("--prompt <text>", "Text prompt describing the audio to generate")
    .option("--length <seconds>", "Duration in seconds")
    .option("--steps <n>", "Number of sampling steps")
    .option("--cfg <value>", "CFG guidance scale")
    .option("--bpm <n>", "Beats per minute")
    .option("--lyrics <text>", "Lyrics text (ace_step)", "")
    .option("--seed <n>", "Random seed for reproducibility")
    .option("--output <path>", "Output file path", "output.wav")
    .option("--models-dir <path>", "Models directory (overrides PYCOMFY_MODELS_DIR)")
    .option("--unet <path>", "UNet model component path (overrides PYCOMFY_ACE_UNET)")
    .option("--vae <path>", "VAE model component path (overrides PYCOMFY_ACE_VAE)")
    .option("--text-encoder-1 <path>", "Text encoder 1 path (overrides PYCOMFY_ACE_TEXT_ENCODER_1)")
    .option("--text-encoder-2 <path>", "Text encoder 2 path (overrides PYCOMFY_ACE_TEXT_ENCODER_2)")
    .option("--async", "Queue job and return a job ID immediately (non-blocking)")
    .addHelpText("after", modelsFooter("create", "audio"))
    .addHelpText("after", buildDefaultsTable("audio"))
    .action(async (opts) => {
      validateModel("create", "audio", opts.model);
      const script = getScript("create", "audio", opts.model);
      if (!script) notImplemented("create", "audio", opts.model);
      const modelsDir = resolveModelsDir(opts.modelsDir);
      const defaults = getModelDefaults("audio", opts.model);
      const audioOpts: AudioOpts = {
        model:        opts.model,
        prompt:       opts.prompt,
        length:       resolveParam(opts.length, defaults?.length, "length"),
        steps:        resolveParam(opts.steps,  defaults?.steps,  "steps"),
        cfg:          resolveParam(opts.cfg,    defaults?.cfg,    "cfg"),
        bpm:          resolveParam(opts.bpm,    defaults?.bpm,    "bpm"),
        lyrics:       opts.lyrics,
        seed:         opts.seed,
        output:       opts.output,
        unet:         opts.unet,
        vae:          opts.vae,
        textEncoder1: opts.textEncoder1,
        textEncoder2: opts.textEncoder2,
      };
      const args = buildAudioArgs(audioOpts, modelsDir);
      if (opts.async) {
        await runAsync("create", "audio", opts.model, script, args, readConfig());
      } else {
        await spawnPipeline(script, args, readConfig());
      }
    });
}

File: cli/commands/create.py
"""``parallax create`` subcommand group — image, video, audio generation."""

from __future__ import annotations

from typing import Annotated, Optional

import typer

from cli._io import resolve_models_dir, save_audio, save_image, save_video_frames
from cli._runners import audio, image, video
from cli.commands._common import ensure_env_on_path

app = typer.Typer(help="Generate media from a text prompt.")


@app.command("image")
def create_image(
    model: Annotated[str, typer.Option("--model", help=f"Model. Choices: {', '.join(image.MODELS)}.")],
    prompt: Annotated[str, typer.Option("--prompt", help="Text prompt describing the image.")],
    negative_prompt: Annotated[Optional[str], typer.Option("--negative-prompt", help="Negative prompt.")] = None,
    width: Annotated[Optional[int], typer.Option("--width", help="Image width in pixels.")] = None,
    height: Annotated[Optional[int], typer.Option("--height", help="Image height in pixels.")] = None,
    steps: Annotated[Optional[int], typer.Option("--steps", help="Sampling steps.")] = None,
    cfg: Annotated[Optional[float], typer.Option("--cfg", help="CFG guidance scale.")] = None,
    seed: Annotated[int, typer.Option("--seed", help="Random seed.")] = 0,
    output: Annotated[str, typer.Option("--output", help="Output file path.")] = "output.png",
    models_dir: Annotated[Optional[str], typer.Option("--models-dir", help="Models directory.")] = None,
    async_mode: Annotated[bool, typer.Option("--async", help="Queue job, return ID immediately.")] = False,
) -> None:
    """Generate an image from a text prompt."""
    if model not in image.MODELS:
        typer.echo(f"Error: unknown model '{model}'. Choices: {', '.join(image.MODELS)}.", err=True)
        raise typer.Exit(code=1)
    if async_mode:
        from cli._async import run_async
        run_async(action="create", media="image", model=model,
                  args={"prompt": prompt, "negative_prompt": negative_prompt, "width": width,
                        "height": height, "steps": steps, "cfg": cfg, "seed": seed,
                        "output": output, "models_dir": models_dir})
        return
    ensure_env_on_path()
    mdir = resolve_models_dir(models_dir)
    w = width  or int(image.default(model, "width",  1024))
    h = height or int(image.default(model, "height", 1024))
    s = steps  or int(image.default(model, "steps",  20))
    c = cfg if cfg is not None else float(image.default(model, "cfg", 7.0))
    try:
        images = image.RUNNERS[model](mdir=mdir, prompt=prompt, neg=negative_prompt, w=w, h=h, s=s, c=c, seed=seed)
        typer.echo(save_image(images, output))
    except (typer.Exit, SystemExit):
        raise
    except Exception as exc:
        typer.echo(f"Error: {exc}", err=True)
        raise typer.Exit(code=1) from exc


_AUDIO_SUPPORTED_MODELS = {"ltx23"}


@app.command("video")
def create_video(
    model: Annotated[str, typer.Option("--model", help=f"Model. Choices: {', '.join(video.MODELS)}.")],
    prompt: Annotated[str, typer.Option("--prompt", help="Text prompt describing the video.")],
    input: Annotated[Optional[str], typer.Option("--input", help="Input image for i2v.")]= None,
    audio: Annotated[Optional[str], typer.Option("--audio", help="Input audio file for audio-conditioned generation (ltx23 only).")] = None,
    width: Annotated[Optional[int], typer.Option("--width", help="Frame width in pixels.")] = None,
    height: Annotated[Optional[int], typer.Option("--height", help="Frame height in pixels.")] = None,
    length: Annotated[Optional[int], typer.Option("--length", help="Number of frames.")] = None,
    fps: Annotated[Optional[int], typer.Option("--fps", help="Output frame rate.")] = None,
    steps: Annotated[Optional[int], typer.Option("--steps", help="Sampling steps.")] = None,
    cfg: Annotated[Optional[float], typer.Option("--cfg", help="CFG guidance scale.")] = None,
    seed: Annotated[int, typer.Option("--seed", help="Random seed.")] = 0,
    output: Annotated[str, typer.Option("--output", help="Output file path.")] = "output.mp4",
    models_dir: Annotated[Optional[str], typer.Option("--models-dir", help="Models directory.")] = None,
    async_mode: Annotated[bool, typer.Option("--async", help="Queue job, return ID immediately.")] = False,
) -> None:
    """Generate a video from a text prompt."""
    from pathlib import Path as _Path

    if model not in video.MODELS:
        typer.echo(f"Error: unknown model '{model}'. Choices: {', '.join(video.MODELS)}.", err=True)
        raise typer.Exit(code=1)

    # --audio validation (before async dispatch so errors surface immediately).
    if audio is not None:
        if model not in _AUDIO_SUPPORTED_MODELS:
            typer.echo("Error: --audio is only supported for model 'ltx23'.", err=True)
            raise typer.Exit(code=1)
        if not _Path(audio).is_file():
            typer.echo(f"Error: audio file not found: {audio}", err=True)
            raise typer.Exit(code=1)
        if input is None:
            typer.echo("Error: --audio requires --input (image).", err=True)
            raise typer.Exit(code=1)

    if async_mode:
        from cli._async import run_async
        run_async(action="create", media="video", model=model,
                  args={"prompt": prompt, "input": input, "audio": audio, "width": width,
                        "height": height, "length": length, "fps": fps, "steps": steps,
                        "cfg": cfg, "seed": seed, "output": output, "models_dir": models_dir})
        return
    if input is not None:
        if not _Path(input).is_file():
            typer.echo(f"Error: input file not found: {input}", err=True)
            raise typer.Exit(code=1)
    ensure_env_on_path()
    mdir = resolve_models_dir(models_dir)
    w = width  or int(video.default(model, "width",  832))
    h = height or int(video.default(model, "height", 480))
    n = length or int(video.default(model, "length", 33))
    f = fps    or int(video.default(model, "fps",    16))
    s = steps  or int(video.default(model, "steps",  20))
    c = cfg if cfg is not None else float(video.default(model, "cfg", 6.0))
    image_obj = None
    if input is not None:
        from PIL import Image as _PIL
        image_obj = _PIL.open(input).convert("RGB")
    try:
        frames = video.RUNNERS[model](mdir=mdir, prompt=prompt, image=image_obj, audio=audio, w=w, h=h, n=n, f=f, s=s, c=c, seed=seed)
        typer.echo(save_video_frames(frames, output, fps=float(f)))
    except (typer.Exit, SystemExit):
        raise
    except Exception as exc:
        typer.echo(f"Error: {exc}", err=True)
        raise typer.Exit(code=1) from exc


@app.command("audio")
def create_audio(
    model: Annotated[str, typer.Option("--model", help=f"Model. Choices: {', '.join(audio.MODELS)}.")],
    prompt: Annotated[str, typer.Option("--prompt", help="Text prompt (genre tags, mood).")],
    length: Annotated[Optional[float], typer.Option("--length", help="Duration in seconds.")] = None,
    steps: Annotated[Optional[int], typer.Option("--steps", help="Sampling steps.")] = None,
    cfg: Annotated[Optional[float], typer.Option("--cfg", help="CFG guidance scale.")] = None,
    bpm: Annotated[Optional[int], typer.Option("--bpm", help="Beats per minute.")] = None,
    lyrics: Annotated[str, typer.Option("--lyrics", help="Lyrics text.")] = "",
    seed: Annotated[int, typer.Option("--seed", help="Random seed.")] = 0,
    output: Annotated[str, typer.Option("--output", help="Output file path.")] = "output.wav",
    models_dir: Annotated[Optional[str], typer.Option("--models-dir", help="Models directory.")] = None,
    async_mode: Annotated[bool, typer.Option("--async", help="Queue job, return ID immediately.")] = False,
) -> None:
    """Generate audio from a text prompt."""
    if model not in audio.MODELS:
        typer.echo(f"Error: unknown model '{model}'. Choices: {', '.join(audio.MODELS)}.", err=True)
        raise typer.Exit(code=1)
    if async_mode:
        from cli._async import run_async
        run_async(action="create", media="audio", model=model,
                  args={"prompt": prompt, "length": length, "steps": steps, "cfg": cfg, "bpm": bpm,
                        "lyrics": lyrics, "seed": seed, "output": output, "models_dir": models_dir})
        return
    ensure_env_on_path()
    mdir = resolve_models_dir(models_dir)
    dur = length if length is not None else float(audio.default(model, "length", 120))
    s   = steps  or int(audio.default(model, "steps", 8))
    c   = cfg    if cfg is not None else float(audio.default(model, "cfg", 1.0))
    b   = bpm    or int(audio.default(model, "bpm", 120))
    try:
        result = audio.RUNNERS[model](mdir=mdir, prompt=prompt, lyrics=lyrics, dur=dur, b=b, s=s, c=c, seed=seed)
        typer.echo(save_audio(result["audio"], output))
    except (typer.Exit, SystemExit):
        raise
    except Exception as exc:
        typer.echo(f"Error: {exc}", err=True)
        raise typer.Exit(code=1) from exc
